### PR TITLE
Allow regex in / search

### LINKED ISSUE #572: Find functionality does not support regex
### What did you do?

In a JS file, from normal mode:
`/\bfor\b.*\bin\b`[Enter]

### What did you expect to happen?

Incremental search should match segments of lines with a for...in loop. e.g. `for (var a in stuff) {`

### What happened instead?

Search opened, all of my keystrokes appeared on the line, but matching did not appear to use regex at all.

Did not match anything at any point after typing `\b`.

Also tried using just `.*` patterns and that didn't work either.

### Technical details:

* VsCodeVim Version: 0.1.4
- VSCode Version: Code - Insiders 1.5.0-insider (1a7ca5a77bf6b63951e499ea93b7316d36118ade, 2016-08-05T07:19:53.669Z)
- OS Version: Windows_NT ia32 10.0.14393

### PR DESCRIPTION
**Pull request by sectioneight**
Fixes #572

### CODE REVIEW

**Comment by johnfn** [comment]
My big concern is I'm not quite sure how well Vim regexp maps to JS regexp. Would you mind doing a little research to see if it is a fairly close mapping or not?

**Comment by sectioneight** [comment]
Sounds good. Hopefully they're close...

**Comment by johnfn** [comment]
I hope so too. Or close enough that you can use a meta regex to translate the normal regex to a js re... oh god, what am I saying?

**Comment by rebornix** [comment]
@sectioneight the regex we use here is the same as Vim Substitution https://github.com/VSCodeVim/Vim/blob/7ac106d8708a1ad9918102abf0e59dbb0609cb4b/src/cmd_line/commands/substitute.ts#L64 . The catch about Regex is no one is following the standard (if there is a standard).

For now I'm using JS's regex in substitution, I'm not quite sure if ppl are satisfied with it but I don't see any complains yet (maybe they don't notice its existence yet).

You may want to write some JS regex to translate Vim's regex to JS's regex ...

**Comment by sectioneight** [comment]
Yes, I've found:

https://github.com/othree/eregex.vim/blob/master/plugin/eregex.vim

To translate Vim regex to PCRE, but of course JS is not PCRE.

There's also https://en.wikipedia.org/wiki/Comparison_of_regular_expression_engines but it's pretty high level.

Sounds like a good first pass would be to use JS regex for search. But I'm open to suggestions.

**Comment by johnfn** [comment]
Yiiiikes. If it's too complicated, let's just do JS regex for now and see if anyone complains. 😉

**Comment by rebornix** [comment]
This PR is good to go IMO.

**Comment by johnfn** [comment]
It still says WIP on it - @sectioneight is this good to go?

**Comment by sectioneight** [comment]
Still need to add a test or two, but it sounds like you guys approve of the approach.

**Comment by johnfn** [comment]
I do; I'm very much a fan of 80% solutions, especially when people very rarely use the other 20%.

**Comment by sectioneight** [comment]
As am I. Will get some tests in tomorrow and ping you for review.

**Comment by sectioneight** [comment]
Ok, ready for review @johnfn

File: src/actions/actions.ts [hunk 1]
```diff
@@ -782,7 +782,7 @@ export class CommandSearchForwards extends BaseCommand {
   isMotion = true;
 
   public async exec(position: Position, vimState: VimState): Promise<VimState> {
-    vimState.searchState = new SearchState(SearchDirection.Forward, vimState.cursorPosition);
+    vimState.searchState = new SearchState(SearchDirection.Forward, vimState.cursorPosition, "", true);
```
Comment (johnfn) [inline src/actions/actions.ts]: Minor, but I'm trying to move towards a style where we explicitly name all our boolean arguments. Something like `{ isRegex: true }` is what I'm thinking here.

**Comment by johnfn** [comment]
Only one small comment.

I think this PR is going to have a big impact for such a small LOC difference! :)

File: src/actions/actions.ts (hunk 1 above)
Comment (sectioneight) [inline src/actions/actions.ts]: Sounds good.

**Comment by sectioneight** [comment]
I'm going to update one more thing on this PR. Right now if you have a partial regex (e.g. `/a(`) it throws an error in the console. I need to try/catch the `SyntaxError` and just do a plain match if it doesn't compile.

File: src/mode/modeHandler.ts [hunk 2]
```diff
@@ -280,10 +295,11 @@ export class SearchState {
     }
   }
 
-  constructor(direction: SearchDirection, startPosition: Position, searchString = "") {
+  constructor(direction: SearchDirection, startPosition: Position, searchString = "", { isRegex = false } = {}) {
```
Comment (sectioneight) [inline src/mode/modeHandler.ts]: I'm not really sure about this syntax (still a TS noob). Is this what you were thinking?

Comment (johnfn) [inline src/mode/modeHandler.ts]: This is actually pure ES6 JS syntax - it's parameter destructuring. I would say to do it without the default object assignment, so just `{ isRegex = false }`

Comment (johnfn) [inline src/mode/modeHandler.ts]: (Which will destructure the object immediately and assign isRegex = false if it can't find a value)

Comment (johnfn) [inline src/mode/modeHandler.ts]: Yeah, that looks fine to me!

**Comment by sectioneight** [comment]
OK, I think we're finally ready to go here

**Comment by sectioneight** [comment]
Ping for review

File: src/mode/modeHandler.ts [hunk 3]
```diff
@@ -214,7 +215,21 @@ export class SearchState {
         ignorecase = false;
       }
 
-      const regex = new RegExp(search.replace(SearchState.specialCharactersRegex, "\\$&"), ignorecase ? 'gi' : 'g');
+      let searchRE = search;
+      if (!this.isRegex) {
+        searchRE = search.replace(SearchState.specialCharactersRegex, "\\$&");
```
Comment (johnfn) [inline src/mode/modeHandler.ts]: This is very cool. I never knew about \$&.

**Comment by johnfn** [comment]
Thanks for the ping. Feel free to ping me on here/slack if I don't review within ~24h. That is almost certainly an indicator that it just slipped through the cracks.

**Comment by johnfn** [comment]
And thanks for the PR, of course :)

**Comment by shaunluttin** [comment]
Does this PR support searching for new lines? If so, how?

**Comment by simahao** [comment]
@shaunluttin ,i have same problem. :(

**Comment by johnfn** [comment]
No, it doesn't support newline search yet. Feel free to open an issue.

**Pull request merged**